browser/tor/tor_launcher_factory.cc

```diff
@@ -127,28 +128,172 @@ void TorLauncherFactory::RemoveObserver(tor::TorProfileServiceImpl* service) {
+                     weak_ptr_factory_.GetWeakPtr()));
+  control_->Subscribe(tor::TorControlEvent::NETWORK_LIVENESS,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STATUS_CLIENT,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STATUS_GENERAL,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STREAM,
+                      base::DoNothing::Once<bool>());
+}
+
+void TorLauncherFactory::GotVersion(bool error, const std::string& version) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (error) {
+    VLOG(1) << "Failed to get version!";
+    return;
+  }
+  VLOG(2) << "Tor version: " << version;
+}
+
+void TorLauncherFactory::GotSOCKSListeners(
+    bool error, const std::vector<std::string>& listeners) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (error) {
+    VLOG(1) << "Failed to get SOCKS listeners!";
+    return;
+  }
+  VLOG(2) << "Tor SOCKS listeners: ";
+  for (auto& listener : listeners) {
+    VLOG(2) << listener;
+  }
+  std::string tor_proxy_uri = kTorProxyScheme + listeners[0];
+  // Remove extra quotes
+  tor_proxy_uri.erase(
+      std::remove(tor_proxy_uri.begin(), tor_proxy_uri.end(), '\"'),
+      tor_proxy_uri.end());
+  for (auto& observer : observers_)
+    observer.NotifyTorNewProxyURI(tor_proxy_uri);
+}
+
+void TorLauncherFactory::OnTorClosed() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  VLOG(2) << "TOR CONTROL: Closed!";
+}
+
+void TorLauncherFactory::OnTorCleanupNeeded(base::ProcessId id) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  VLOG(2) << "Killing old tor process pid=" << id;
+  // Dispatch to launcher thread
+  content::GetProcessLauncherTaskRunner()
+    ->PostTask(FROM_HERE,
+               base::BindOnce(&TorLauncherFactory::KillOldTorProcess,
+                              base::Unretained(this),
+                              std::move(id)));
+}
+
+
+void TorLauncherFactory::KillOldTorProcess(base::ProcessId id) {
+  DCHECK(content::CurrentlyOnProcessLauncherTaskRunner());
+  base::Process tor_process = base::Process::Open(id);
+  if (tor_process.IsValid())
+    tor_process.Terminate(0, false);
+}
+
+void TorLauncherFactory::RelaunchTor() {
+  Init();
+  control_->PreStartCheck(
+      config_.tor_watch_path(),
+      base::BindOnce(&TorLauncherFactory::OnTorControlCheckComplete,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void TorLauncherFactory::OnTorEvent(
+    tor::TorControlEvent event, const std::string& initial,
+    const std::map<std::string, std::string>& extra) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  VLOG(3) << "TOR CONTROL: event "
+             << (*tor::kTorControlEventByEnum.find(event)).second
+             << ": " << initial;
+  if (event == tor::TorControlEvent::STATUS_CLIENT) {
+    if (initial.find("BOOTSTRAP") != std::string::npos) {
+      const char prefix[] = "PROGRESS=";
```

Comment:
it would be better to define these consts (and others that should be const like "CIRCUIT_ESTABLISHED") together at the top of the file

Reply:
fixed in 12c003763561ef5dc1ef014d4769d24a84c58883